[core] Refactor `<Account />`

bharatkashyap opened the pull request:
-  Fixes #3971 
- Adds `slots` for `signInButton` and `signOutButton` and `menuItems`
- Fixes an `anchorEl` issue with the previous `<Account />` 
- Adds a custom menu demo to the docs

https://deploy-preview-3992--mui-toolpad-docs.netlify.app/toolpad/core/react-account/

Janpot commented:
This component could be much more than a simple menu. It feels like a regression to me.

![Screenshot 2024-08-27 at 10 32 24](https://github.com/user-attachments/assets/faab6e47-54d8-480b-9221-6d19ae771f6c)
to
![Screenshot 2024-08-27 at 10 32 29](https://github.com/user-attachments/assets/347dd7b9-34e1-40bf-98c7-19d71511bdc7)

bharatkashyap commented:
> This component could be much more than a simple menu. It feels like a regression to me.
> 
> ![Screenshot 2024-08-27 at 10 32 24](https://private-user-images.githubusercontent.com/2109932/361720822-faab6e47-54d8-480b-9221-6d19ae771f6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NjE4MDMsIm5iZiI6MTcyNDc2MTUwMywicGF0aCI6Ii8yMTA5OTMyLzM2MTcyMDgyMi1mYWFiNmU0Ny01NGQ4LTQ4MGItOTIyMS02ZDE5YWU3NzFmNmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjdUMTIyNTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YzNlZjE5YzdiZjc4NzAxOTUxMjQ3MmJmYzlhYjNmZmM2YmQ3MjlmMGY4YTFiYzVjNjdhYjRhMzc2OTZkODMyZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5bX3fGIeBueG0rK3A6p3MerqIbeCXPOvWEky04j5flc) to ![Screenshot 2024-08-27 at 10 32 29](https://private-user-images.githubusercontent.com/2109932/361720686-347dd7b9-34e1-40bf-98c7-19d71511bdc7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NjE4MDMsIm5iZiI6MTcyNDc2MTUwMywicGF0aCI6Ii8yMTA5OTMyLzM2MTcyMDY4Ni0zNDdkZDdiOS0zNGUxLTQwYmYtOThjNy0xOWQ3MTUxMWJkYzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjdUMTIyNTAzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjkwYjc4YzNiODEzNWJkOTVjNThiZjY3OGZkNTJjOWUxOTc2OTk2N2E1OTRmNzdjMTg0Y2EzOGFkMGRlYTY5MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.j6gVMOvI4w0MYyXjfLV6qeb9xudROfhZQ3oKRyp4dss)

I'll explore continuing to use the `Popover`, but I feel like the account/user button component is almost always going to be a list of options and text visually separated by dividers and icons on a temporary surface. Seems apt to be a `Menu`.

prakhargupta1 commented:
I checked two instances of it being a menu (my best guess) component. Since, it's already done I would suggest that we continue with it. In the future, we can revisit it.  

Zendesk:
<img width="341" alt="Screenshot 2024-08-29 at 1 21 13 PM" src="https://github.com/user-attachments/assets/59133ce9-0b6f-4ae4-8659-ce52fd97420a">

HiBob:
<img width="225" alt="Screenshot 2024-08-29 at 1 20 54 PM" src="https://github.com/user-attachments/assets/6711a2c0-b01a-46cb-b0a6-d5ce5ddf314b">

Janpot commented:
Nothing should prevent you from just rendering a [`MenuList`](https://mui.com/material-ui/react-menu/#icon-menu) in the slot. Why would we restrict every usage of this component to be this one?

![Screenshot 2024-08-28 at 15 18 31](https://github.com/user-attachments/assets/d2714eed-12a5-47fe-8744-12366ec51396)

Janpot commented:
Nothing should prevent you from just rendering a [MenuList](https://mui.com/material-ui/react-menu/#icon-menu) in the slot. Why would we restrict every usage of this component to be this one?

![Screenshot 2024-08-28 at 15 18 31](https://github.com/user-attachments/assets/7e09253c-af30-4351-a09e-5c56d50b73a8)

prakhargupta1 commented:
> Nothing should prevent you from just rendering a [MenuList](https://mui.com/material-ui/react-menu/#icon-menu) in the slot. Why would we restrict every usage of this component to be this one?

Ah ok, so I guess it's about whether to give the option of a slot as menu item versus changing it to a popover component that can show any UI (like the Google UI you shared) through a slot. 
I still think, it could be done in the next iteration, based on user demand.

Janpot commented:
> Ah ok, so I guess it's about whether to give the option of a slot as menu item versus changing it to a popover component that can show any UI (like the Google UI you shared) through a slot.

No, it's already a popover, this PR is changing it to a menu

prakhargupta1 commented:
> No, it's already a popover, this PR is changing it to a menu

Oh ok, in that case, I agree, a popover would be a more versatile choice.

bharatkashyap commented:
Updated to use `Popover` instead of `Menu` by default

Janpot commented on packages/toolpad-core/src/Account/Account.tsx:
```diff
@@ -47,109 +65,140 @@ export interface AccountProps {
-  const { slotProps, signInLabel = 'Sign In', signOutLabel = 'Sign Out' } = props;
-  const session = React.useContext(SessionContext);
-
+  const { slots, slotProps, signInLabel = 'Sign In', signOutLabel = 'Sign Out' } = props;
```
How does X handle customizing their labels in e.g. the datagrid?

bharatkashyap replied:
The closest example I can find in the grid is customising the pagination labels https://mui.com/x/react-data-grid/pagination/#page-size-options

Janpot replied:
https://mui.com/x/react-data-grid/localization/#locale-text

bharatkashyap replied:
Do you mean we should drop the `signInLabel` and `signOutLabel` props in favour of a `LabelsProvider` of some sort?

Janpot replied:
The grid supports a `localeText` property. We could opt for following the same convention? It also reads those from the theme. We can add support for that later.

Janpot commented on packages/toolpad-core/src/Account/Account.tsx:
```diff
@@ -47,109 +65,140 @@ export interface AccountProps {
-      >
-        <SessionAvatar session={session} sx={{ width: 32, height: 32, ...slotProps?.avatar }} />
-      </IconButton>
+      <Box sx={{ display: 'flex', alignItems: 'center', textAlign: 'center' }}>
```
We could consider starting adopting the habit of avoiding `Box` components?
```
styled('div'){{ display: 'flex', alignItems: 'center', textAlign: 'center' }}
```

Janpot commented on docs/data/toolpad/core/components/account/account.md:
```diff
@@ -12,18 +12,20 @@ The `Account` component is a quick and easy way to display an account management
 
 `Account` can take different labels for the sign in and sign out buttons through the `signInLabel` and `signOutLabel` props. Deeper changes can be made by passing in `slotProps` to the underlying components.
 
-{{"demo": "AccountCustom.js", "bg": "gradient" }}
```
Pure visually, I wonder if it makes sense to show those demos inside of an empty `AppBar`.

apedroferreira commented on packages/toolpad-core/src/Account/Account.tsx:
```diff
@@ -1,39 +1,67 @@
-   * @example { signInButton: { color: 'primary' }, signOutButton: { color: 'secondary' } }
+   * The components used for each slot inside.
+   */
+  slots?: {
```
You need to export the Slots in their own interface to automatically generate the documentation for them.
Take a look at https://github.com/mui/mui-toolpad/blob/master/packages/toolpad-core/src/DashboardLayout/DashboardLayout.tsx, line 253, should be a simple change!

apedroferreira commented on docs/pages/toolpad/core/api/account.json:
```diff
@@ -1,13 +1,21 @@
-    "signOutLabel": { "type": { "name": "string" }, "default": "'Sign Out'" },
+    "localeText": {
+      "type": { "name": "shape", "description": "{ signInLabel: string, signOutLabel: string }" },
+      "default": "DEFAULT_LABELS"
```
Does MUI X also list constants as defaults like this, even if they're not accessible for users?
Just curious, as not sure if there's a better way, but if it's following an existing pattern I guess it's fine!

bharatkashyap replied:
Yep: https://github.com/mui/mui-x/blob/master/packages/x-data-grid/src/constants/localeTextConstants.ts

apedroferreira reviewed (changes requested):
Looks cool! Got some feedback mostly on docs/design, nothing major except generating the slots documentation automatically by exporting a type interface!

### Some design feedback on this demo:

<img width="905" alt="Screenshot 2024-09-06 at 18 51 14" src="https://github.com/user-attachments/assets/bbfad7c9-2f9c-4608-ac96-c7178dc81049">

- The items aligned to the center don't look very good, it's better if they're aligned to the left.
- There's too much space between the picture and the title and description, I think you can decrease the margin a bit! Not that important though if this is a default.
- A Popover that opens another popover doesn't seem great, I guess a single list should work and be simpler too.

### Some feedback on the new slots:

On a first read I'm not exactly sure what the `menuItems` slot replaces. Is it everything inside the Popover? Just the middle area? Maybe we could make it a bit clearer, even if just with the text in the docs.

Also, we're using `localeText` inside `slotProps` to customize labels, right? The current version of the preview docs seems not to be updated with that.

bharatkashyap commented:
> Looks cool! Got some feedback mostly on docs/design, nothing major except generating the slots documentation automatically by exporting a type interface!

Agree with almost everything here and addressed it!

> A Popover that opens another popover doesn't seem great, I guess a single list should work and be simpler too.

Added this to demonstrate how one would be able to create such a menu if they want to. It's not an uncommon UX pattern, and I feel it should be okay to include as an example in the docs? Let me know what you think about that.

apedroferreira commented on docs/data/toolpad/core/components/account/account.md:
```diff
@@ -26,12 +26,16 @@ When signed out, the component renders as an inline sign in button within the da
-`Account` can take different labels for the sign in and sign out buttons through the `signInLabel` and `signOutLabel` props. Deeper changes can be made by passing in `slotProps` to the underlying components.
+The underlying `signInButton`, `signOutButton` and `iconButton` components can be customized by passing in `slotProps` to the `Account` component.
+
+Labels for the sign in and sign out buttons through the `localeText` prop.
```
I think this sentence about the labels is missing something ("can be customized?").
